Move and revise overview for Security section (#43176)
kubernetes/website: content/en/docs/concepts/security/_index.md

weight: 85
description: >
  Concepts for keeping your cloud-native workload secure.
simple_list: true
---

This section of the Kubernetes documentation aims to help you learn to run
workloads more securely, and about the essential aspects of keeping a
Kubernetes cluster secure.

Kubernetes is based on a cloud-native architecture, and draws on advice from the
{{< glossary_tooltip text="CNCF" term_id="cncf" >}} about good practice for
cloud native information security.

Read [Cloud Native Security and Kubernetes](/docs/concepts/security/cloud-native-security/)
for the broader context about how to secure your cluster and the applications that
you're running on it.

## Kubernetes security mechanisms {#security-mechanisms}

Kubernetes includes several APIs and security controls, as well as ways to
define [policies](#policies) that can form part of how you manage information security.

### Control plane protection

> A key security mechanism for any Kubernetes is to

Reviewer: ```suggestion
A key security mechanism for any Kubernetes cluster is to
```
Nit

[control access to the Kubernetes API](/docs/concepts/security/controlling-access).

> You can define [encryption at rest](/docs/tasks/administer-cluster/encrypt-data/)

Reviewer: Nit: Recommend calling out TLS first in this paragraph, since it is essential to virtually any effort to secure a kubernetes cluster. Then move to "You can also configure encryption at rest..." because it's a good idea but less critical for many common use cases.

Author: (reworded)

for the Kubernetes control plane. By default, Kubernetes also expects you to use
TLS to provide [data encryption in transit](/docs/tasks/tls/managing-tls-in-a-cluster/)
within the control plane, and between the control plane and its clients.

### Secrets

The [Secret](/docs/concepts/configuration/secret/) API provides basic protection for

> configuration values that require confidentiality.

Reviewer 2: Should we add that secrets are only base64 encoded and are stored unencrypted in the API server? (A premonition to the caution block under the secrets documentation)

Author: Maybe you'd like to propose that for a follow-up PR?

* [Secrets in Kubernetes](/docs/concepts/configuration/secret/)
* [Pod security standards](/docs/concepts/security/pod-security-standards/)
* [RuntimeClasses](/docs/concepts/containers/runtime-class)

Learn the context:

* CNCF [white paper](https://www.cncf.io/wp-content/uploads/2022/06/CNCF_cloud-native-security-whitepaper-May2022-v2.pdf)
  on cloud native security.
* CNCF [white paper](https://github.com/cncf/tag-security/blob/f80844baaea22a358f5b20dca52cd6f72a32b066/supply-chain-security/supply-chain-security-paper/CNCF_SSCP_v1.pdf)
  on good practices for securing a software supply chain.
* [Fixing the Kubernetes clusterf\*\*k: Understanding security from the kernel up](https://archive.fosdem.org/2020/schedule/event/kubernetes/) (FOSDEM 2020)
* [Kubernetes Security Best Practices](https://www.youtube.com/watch?v=wqsUfvRyYpw) (Kubernetes Forum Seoul 2019)
* [Towards Measured Boot Out of the Box](https://www.youtube.com/watch?v=EzSkU3Oecuw) (Linux Security Summit 2016)

Get certified:

* [Certified Kubernetes Security Specialist](https://training.linuxfoundation.org/certification/certified-kubernetes-security-specialist/)
  certification and official training course.

> Read more in this section:

Author: This relies on `simple_list: true` in the front matter. See the [preview](https://deploy-preview-43176--kubernetes-io-main-staging.netlify.app/docs/concepts/security/).

  Concepts for keeping your cloud-native workload secure.
simple_list: true
---

This section of the Kubernetes documentation aims to help you learn to run
workloads more securely, and about the essential aspects of keeping a
Kubernetes cluster secure.

Kubernetes is based on a cloud-native architecture, and draws on advice from the
{{< glossary_tooltip text="CNCF" term_id="cncf" >}} about good practice for
cloud native information security.

## Cloud native information security

The CNCF [white paper](https://www.cncf.io/wp-content/uploads/2022/06/CNCF_cloud-native-security-whitepaper-May2022-v2.pdf)
on cloud native security defines security controls and practices that are
appropriate to different _lifecycle phases_.

## _Develop_ lifecycle phase {#lifecycle-phase-develop}

- Ensure the integrity of development environments.
- Design applications following good practice for information security,
  appropriate for your context.
- Consider end user security as part of solution design.

To achieve this, you can:

1. Adopt an architecture, such as [zero trust](https://glossary.cncf.io/zero-trust-architecture/),
   that minimises attack surfaces, even for internal threats.

> 1. Define a code review process that considers security concerns.

Author: I don't think so. Reviewing docs isn't really relevant to securing a cluster.

> 1. Define a code review process that considers security concerns.

Reviewer 3: +1 to what Tim said.

provide and manage security identities for workloads and cluster
components.

Kubernetes uses TLS to protect API traffic; make sure to deploy the cluster using
TLS (including for traffic between nodes and the control plane), and protect the
encryption keys. If you use Kubernetes own API for
[CertificateSigningRequests](/docs/reference/access-authn-authz/certificate-signing-requests/#certificate-signing-requests),
pay special attention to restricting misuse there.

### Runtime protection: compute {#protection-runtime-compute}

{{< glossary_tooltip text="Containers" term_id="container" >}} provide two
things: isolation between different applications, and a mechanism to combine
those isolated applications to run on the same host computer. Those two
aspects, isolation and aggregation, mean that runtime security involves
trade-offs and finding an appropriate balance.

Kubernetes relies on a {{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}
to actually set up and run containers. The Kubernetes project does
not recommend a specific container runtime and you should make sure that

> the runtime(s) that you choose meet your information security needs.

Author: No, because the glossary tooltip does that automatically.

1. Authenticate connections between cluster nodes and any network storage they rely
   upon.
1. Implement data encryption within your own application.

For encryption keys, generating these within specialized hardware provides
the best protection against disclosure risks. A _hardware security module_
can let you perform cryptographic operations without allowing the security
key to be copied elsewhere.

### Networking and security

You should also consider network security measures, such as
[NetworkPolicy](/docs/concepts/services-networking/network-policies/) or a
[service mesh](https://glossary.cncf.io/service-mesh/).
Some network plugins for Kubernetes provide encryption for your
cluster network, using technologies such as a virtual
private network (VPN) overlay.
By design, Kubernetes lets you use your own networking plugin for your
cluster (if you use managed Kubernetes, the person or organization
managing your cluster may have chosen a network plugin for you).

The network plugin you choose and the way you integrate it can have a
strong impact on the security of information in transit.

### Observbility and runtime security

> As your applications run they - and the cluster they run on -  give you ways to

Author: How's this now?

weight: 85
description: >
  Concepts for keeping your cloud-native workload secure.
simple_list: true
---

This section of the Kubernetes documentation aims to help you learn to run
workloads more securely, and about the essential aspects of keeping a
Kubernetes cluster secure.

Kubernetes is based on a cloud-native architecture, and draws on advice from the

> {{< glossary_tooltip text="CNCF" term_id="cncf" >}} about good practice for

Reviewer 4: ```suggestion
{{< glossary_tooltip text="CNCF" term_id="cncf" >}} about good practices for
```

Author: I could make this change, but: I'm also happy with the original wording. It doesn't feel like this is a blocker to merging the PR.

- Ensure the security of the supply chain for container images you execute.
- Ensure the security of the supply chain for the cluster and other components
  that execute your application. An example of another component might be an
  external database that your cloud-native application uses for persistence.

To achieve this, you can:

1. Scan container images and other artifacts for known vulnerabilities.
1. Ensure that software distribution uses encryption in transit, with
   a chain of trust for the software source.
1. Adopt and follow processes to update dependencies when updates are
   available, especially in response to security announcements.
1. Use validation mechanisms such as digital certificates for supply
   chain assurance.
1. Subscribe to feeds and other mechanisms to alert you to security
   risks.
1. Restrict access to artifacts. Place container images in a
   [private registry](/docs/concepts/containers/images/#using-a-private-registry)
   that only allows authorized clients to pull images.

## _Deploy_ lifecycle phase {#lifecycle-phase-deploy}

Ensure appropriate restrictions on what can be deployed, who can deploy it,

> and where it can be deployed to.

Reviewer 4: ```suggestion
and where it can be deployed.
```

Author: I (still) think “where is can be deployed to” is how people who speak English would say it.

Reviewer 3: +1 to what Tim said.

You can enforce measures from the _distribute_ phase, such as verifying the
cryptographic identity of container image artefacts.

When you deploy Kubernetes, you also set the foundation for your
applications' runtime environment: a Kubernetes cluster (or
multiple clusters).
That IT infrastructure must provide the security guarantees that higher
layers expect.

## _Runtime_ lifecycle phase {#lifecycle-phase-runtime}

> The Runtime phase comprises three critical areas: [compute](#protection-runtime-compute),

Reviewer 4: ```suggestion
The Runtime phase comprises of three critical areas: [compute](#protection-runtime-compute),
```

Author: I'm afraid I won't accept this suggestion as it doesn't look like idiomatic English grammar.

[access](#protection-runtime-access), and [storage](#protection-runtime-storage).

### Runtime protection: access {#protection-runtime-access}

> The Kubernetes API is what makes your cluster work. Protecting this API is key
> to providing effective cluster security.

Reviewer 4: ```suggestion
The Kubernetes API is what makes your cluster work. Protecting this API is a key
for providing effective cluster security.
```

Author: I'm afraid I won't accept this suggestion as it doesn't look like idiomatic English grammar.

Other pages in the Kubernetes documentation have more detail about how to set up
specific aspects of access control. The [security checklist](/docs/concepts/security/security-checklist/)
has a set of suggested basic checks for your cluster.

Beyond that, securing your cluster means implementing effective
[authentication](/docs/concepts/security/controlling-access/#authentication) and
[authorization](/docs/concepts/security/controlling-access/#authorization) for API access. Use [ServiceAccounts](/docs/concepts/security/service-accounts/) to
provide and manage security identities for workloads and cluster
components.

Kubernetes uses TLS to protect API traffic; make sure to deploy the cluster using

> TLS (including for traffic between nodes and the control plane), and protect the

Reviewer 4: ```suggestion
TLS (including the traffic between nodes and control plane), and protect the
```

Author: I wouldn't feel happy putting my name to that sentence.

encryption keys. If you use Kubernetes' own API for
[CertificateSigningRequests](/docs/reference/access-authn-authz/certificate-signing-requests/#certificate-signing-requests),

> pay special attention to restricting misuse there.

Reviewer 4: ```suggestion
pay special attention for restricting misuse there.
```

Author: I'm afraid I prefer my text. Was there a reason you proposed an edit there?

Reviewer 5: I agree with @sftim here, the original wording is more grammatically correct.

### Runtime protection: compute {#protection-runtime-compute}

{{< glossary_tooltip text="Containers" term_id="container" >}} provide two
things: isolation between different applications, and a mechanism to combine
those isolated applications to run on the same host computer. Those two
aspects, isolation and aggregation, mean that runtime security involves

> trade-offs and finding an appropriate balance.

Reviewer 4: ```suggestion
trade-offs in finding an appropriate balance.
```

Author: I'm afraid I prefer my text.

Kubernetes relies on a {{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}
to actually set up and run containers. The Kubernetes project does
not recommend a specific container runtime and you should make sure that
the runtime(s) that you choose meet your information security needs.

To protect your compute at runtime, you can:

1. Enforce [Pod security standards](/docs/concepts/security/pod-security-standards/)
   for applications, based on the level of trust you have in them.
1. Run a specialized operating system on your nodes that is designed specifically
   for running containerized workloads. This is typically based on a read-only
   operating system (_immutable image_) that provides only the services
   essential for running containers.

   Container-specific operating systems help to isolate system components and
   present a reduced attack surface in case of a container escape.
1. Define [ResourceQuotas](/docs/concepts/policy/resource-quotas/) to
   fairly allocate shared resources, and use

>    mechanisms such as [LimitRanges](/docs/concepts/policy/limit-range/)

Reviewer 4: ```suggestion
   mechanisms such as [LimitRange](/docs/concepts/policy/limit-range/)
```

Author: I'm afraid I won't accept this suggestion as it would look odd to have ResourceQuotas in plural and LimitRange in singular.

   to ensure that Pods specify their resource requirements.
1. Partition workloads across different nodes.
   Use [node isolation](/docs/concepts/scheduling-eviction/assign-pod-node/#node-isolation-restriction)
   mechanisms, either from Kubernetes itself or from the ecosystem, to ensure that

>    Pods with different trust contexts are run on separate sets of nodes.

Reviewer 4: ```suggestion
   the Pods with different trust contexts will run on separate sets of node.
```

Author: I'm afraid I won't accept this suggestion as it doesn't look like idiomatic English grammar.

1. Use a {{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}
   that provides security restrictions.
1. On Linux nodes, use a Linux security module such as [AppArmor](/docs/tutorials/security/apparmor/) (beta)
   or [seccomp](/docs/tutorials/security/seccomp/).

### Runtime protection: storage {#protection-runtime-storage}

To protect storage for your cluster and the applications that run there, you can:

1. Integrate your cluster with an external storage plugin that provides encryption at
   rest for volumes.
1. Enable [encryption at rest](/docs/tasks/administer-cluster/encrypt-data/) for
   API objects.

> 1. Protect data durability using backups. Verify you can restore these if you need to.

Reviewer 4: ```suggestion
1. Protect data durability using backups. Verify that you can restore these, whenever you need to.
```

1. Authenticate connections between cluster nodes and any network storage they rely

>    upon.

Reviewer 4: ```suggestion
   upon it.
```

Author: I'm afraid I won't accept this suggestion as it doesn't look like typical English grammar.

workloads more securely, and about the essential aspects of keeping a
Kubernetes cluster secure.

Kubernetes is based on a cloud-native architecture, and draws on advice from the
{{< glossary_tooltip text="CNCF" term_id="cncf" >}} about good practice for
cloud native information security.

Read on through this page for an overview of how Kubernetes is designed to
help you deploy a secure cloud native platform, or jump to the
[further reading](#what-s-next) if you are looking for a specific topic.

## Cloud native information security

The CNCF [white paper](https://www.cncf.io/wp-content/uploads/2022/06/CNCF_cloud-native-security-whitepaper-May2022-v2.pdf)
on cloud native security defines security controls and practices that are
appropriate to different _lifecycle phases_.

## _Develop_ lifecycle phase {#lifecycle-phase-develop}

- Ensure the integrity of development environments.
- Design applications following good practice for information security,
  appropriate for your context.
- Consider end user security as part of solution design.

To achieve this, you can:

1. Adopt an architecture, such as [zero trust](https://glossary.cncf.io/zero-trust-architecture/),

>    that minimises attack surfaces, even for internal threats.

Reviewer 6: minimizes

1. Use validation mechanisms such as digital certificates for supply
   chain assurance.
1. Subscribe to feeds and other mechanisms to alert you to security
   risks.
1. Restrict access to artifacts. Place container images in a
   [private registry](/docs/concepts/containers/images/#using-a-private-registry)
   that only allows authorized clients to pull images.

## _Deploy_ lifecycle phase {#lifecycle-phase-deploy}

Ensure appropriate restrictions on what can be deployed, who can deploy it,
and where it can be deployed to.
You can enforce measures from the _distribute_ phase, such as verifying the
cryptographic identity of container image artefacts.

When you deploy Kubernetes, you also set the foundation for your
applications' runtime environment: a Kubernetes cluster (or
multiple clusters).
That IT infrastructure must provide the security guarantees that higher
layers expect.

## _Runtime_ lifecycle phase {#lifecycle-phase-runtime}

The Runtime phase comprises three critical areas: [compute](#protection-runtime-compute),
[access](#protection-runtime-access), and [storage](#protection-runtime-storage).

### Runtime protection: access {#protection-runtime-access}

> The Kubernetes API is what makes your cluster work. Protecting this API is key

Reviewer 6: nit: This sentence is a bit general.

1. Authenticate connections between cluster nodes and any network storage they rely
   upon.
1. Implement data encryption within your own application.

For encryption keys, generating these within specialized hardware provides
the best protection against disclosure risks. A _hardware security module_
can let you perform cryptographic operations without allowing the security
key to be copied elsewhere.

### Networking and security

You should also consider network security measures, such as
[NetworkPolicy](/docs/concepts/services-networking/network-policies/) or a
[service mesh](https://glossary.cncf.io/service-mesh/).
Some network plugins for Kubernetes provide encryption for your
cluster network, using technologies such as a virtual
private network (VPN) overlay.
By design, Kubernetes lets you use your own networking plugin for your
cluster (if you use managed Kubernetes, the person or organization
managing your cluster may have chosen a network plugin for you).

The network plugin you choose and the way you integrate it can have a
strong impact on the security of information in transit.

> ### Observbility and runtime security

Reviewer 6: Can you fix the spelling error in the heading?

Author: Sure!

ecosystem around Kubernetes. Kubernetes provides extension mechanisms
to let those ecosystem projects implement their own policy controls
on source code review, container image approval, API access controls,
networking, and more.

For more information about policy mechanisms and Kubernetes,
read [Policies](/docs/concepts/policy/).

## {{% heading "whatsnext" %}}

Learn about related Kubernetes security topics:

* [Securing your cluster](/docs/tasks/administer-cluster/securing-a-cluster/)
* [Known vulnerabilities](/docs/reference/issues-security/official-cve-feed/)
  in Kubernetes (and links to further information)
* [Data encryption in transit](/docs/tasks/tls/managing-tls-in-a-cluster/) for the control plane
* [Data encryption at rest](/docs/tasks/administer-cluster/encrypt-data/)
* [Controlling Access to the Kubernetes API](/docs/concepts/security/controlling-access)
* [Network policies](/docs/concepts/services-networking/network-policies/) for Pods
* [Secrets in Kubernetes](/docs/concepts/configuration/secret/)
* [Pod security standards](/docs/concepts/security/pod-security-standards/)
* [RuntimeClasses](/docs/concepts/containers/runtime-class)

> Learn the context:

Reviewer 6: I suggest to remove these links and the link to the certification course (seems like advertising).

Author: I think we'd quite like to advertise the course - it's how we (our foundation, the CNCF) make some of the revenue that pays for running the website.
